Server/App/api/library.py

```diff
@@ -1,56 +1,38 @@
-import json
 import logging
 
 from config import StatusCode
-from utils.serializer import serializer
-from utils.identity import generate_uuid
+from model.deck import Deck
+from model.library import Library
+from policy.deck import DeckPolicy
 from utils.base_handler import BaseHandler
-from utils.wrappers import authorize_request_and_create_db_connector, extract_user_id, require_params
+from utils.wrappers import authorize_request_and_create_db_connector, extract_user_id, \
+    require_params, optional_params
 
 
 class LibraryHandler(BaseHandler):
     @authorize_request_and_create_db_connector
     @extract_user_id
     def get(self, user_id, connector):
         """Return metadata for all the decks in a user's library."""
-        
-        # Retrieve user's.
-        library = connector.call_procedure('FETCH_LIBRARY', user_id)
 
-        for deck in library:
-            deck['tags'] = map(lambda x: x['tag'], connector.call_procedure('GET_TAGS', deck["uuid"]))
-            
-
-        # Return library to the user.
-        return self.make_response(response=json.dumps(library, default=serializer),
-            status=StatusCode.OK)
+        return self.make_response(response=Library(user_id, connector=connector).to_json())
 
 
 class LibraryAddHandler(BaseHandler):
     @authorize_request_and_create_db_connector
     @require_params('uuid')
+    @optional_params('share_code')
     @extract_user_id
-    def post(self, user_id, uuid, connector):
+    def post(self, user_id, share_code, uuid, connector):
         """Add a remote deck to a user's library."""
 
-        try:
-            connector.begin_transaction()
-
-            results = connector.call_procedure('FETCH_DECK', user_id, uuid)
-            if len(results) != 1:
-                return self.make_response(status=StatusCode.NOT_FOUND)
-
-            deck = results[0]
-
-            if (not deck['public'] and deck['owner'] != user_id):
-                return self.make_response(status=StatusCode.UNAUTHORIZED)
-
-            connector.call_procedure('LIBRARY_ADD', user_id, uuid)
-
-            connector.end_transaction()
+        deck = DeckPolicy.can_view(uuid, user_id, share_code, connector=connector)
+        if deck is None:
+            return self.make_response(status=StatusCode.NOT_FOUND)
 
+        try:
+            connector.call_procedure_transactionally('LIBRARY_ADD', user_id, deck.uuid)
```

Comment:
Might as well make this a function of the Library model to keep things consistent?